Add static analysis GitHub action

martinkennelly opened the pull request:
Add static analysis GitHub action.

Add golangci-lint action with extra linters besides the default set [1].
Add hadolint and shellcheck.

[1] https://golangci-lint.run/usage/linters

zshi-redhat commented:
Good work, looks like it captures many "Errors".
@martinkennelly would you want to fix those "Errors" in this PR or a separate one?

martinkennelly commented:
Separate. I will start to work on them. I am hoping, we can add these basic actions to all our repos. In the last community meeting, we discussed trialing them in NRI first.
I am currently working on e2e test action as well with kind.

zshi-redhat commented:
> Separate. I will start to work on them. I am hoping, we can add these basic actions to all our repos. In the last community meeting, we discussed trialing them in NRI first.
> I am currently working on e2e test action as well with kind.

@martinkennelly shall we merge this PR or wait until the errors be fixed.

martinkennelly commented:
> > Separate. I will start to work on them. I am hoping, we can add these basic actions to all our repos. In the last community meeting, we discussed trialing them in NRI first.
> > I am currently working on e2e test action as well with kind.
> 
> @martinkennelly shall we merge this PR or wait until the errors be fixed.

@zshi-redhat merge please. @emmakenny is working on this and will submit separate PR.

martinkennelly commented:
Forced pushed changes: Add golangci lint config file, created a make target, rebased to master and executed go mod tidy.

martinkennelly commented:
@zshi-redhat I pushed the fixes here for static analysis. I fixed most and left a few that needs to be addressed in another PR. Please see the commit messages for more details. All the commits address static scan issues but I broke some fixes out into separate commits to make it easier for you to review.

I also rebased this entire patch series ontop of the e2e features tests (#98) for regression testing and got an expected outcome including self hosted test to validate hugepages downward API info feature - output:
https://github.com/martinkennelly/network-resources-injector/runs/2750114201?check_suite_focus=true
